dt/si: Dial timeouts way down in fast uploads mode

oleiman opened the pull request:
In at least one ducktape test, we can get into a situation where a leadership transfer followed by a faulty volley of segment uploads can eat away most of our time budget for the test. To account for that, simply dial down the upload timeouts, operating under the assumption that fast uploads should result in smaller segments.

Also introduces `rptest.util.wait_until_with_progress_check`, which allows us to set a longer top-level timeout that short circuits if some specified value stops changing.

Ran the problematic test config against this change a whole bunch of times (3 x 350 x 2) w/ no failures:
- [ducktape-debug](https://buildkite.com/redpanda/redpanda/builds/69260#01981c35-2727-482f-89c2-a93c4477c6a8)
- [ducktape-release](https://buildkite.com/redpanda/redpanda/builds/69260#01981c35-272f-4cdc-9173-83467295bb70)

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v25.2.x
- [x] v25.1.x
- [x] v24.3.x
- [ ] v24.2.x

## Release Notes

* none

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

oleiman committed:
dt/si: Dial timeouts way down in fast uploads mode

In at least one ducktape test, we can get into a situation where a leadership
transfer followed by a faulty volley of segment uploads can eat away most of
our time budget for the test. In fast_upload mode, we expect segments on the
smaller side, and it's convenient to let them fail quickly back to be retried.

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman commented:
/ci-repeat 3
skip-units
dt-repeat=20
tests/rptest/tests/archive_retention_test.py::ClusterArchiveRetentionTest.test_delete

oleiman commented:
/ci-repeat 3
skip-redpanda-build
skip-units
dt-repeat=20
tests/rptest/tests/archive_retention_test.py::CloudArchiveRetentionTest.test_delete

oleiman commented:
/ci-repeat 3
skip-redpanda-build
skip-units
dt-repeat=50
tests/rptest/tests/archive_retention_test.py::CloudArchiveRetentionTest.test_delete

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--d27da4f1-5f6d-4083-830d-a62c25d48ff7-->
<!--TEST RESULTS-->
<!--92170796-d979-4aba-9cde-1bd2a356dcc2-->
<!--TEST RESULTS-->
<!--6ff295e7-a94b-4114-9361-1a6c7a21be36-->
<!--TEST RESULTS-->
<!--c4ecd86a-b7ac-453e-b749-52e5127a995d-->
<!--TEST RESULTS-->
<!--3bae6fd9-9df1-49d8-bb75-af38d028678a-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69220">build#69220</a></summary>


| test_class                | test_method   | test_arguments                                              | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                    |
|:--------------------------|:--------------|:------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------|
| CloudArchiveRetentionTest | test_delete   | {"cloud_storage_type": 1, "retention_type": "retention.ms"} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69220#019819f0-dba3-4931-a7e5-969f7c3b0d4e | FLAKY         | 69/70    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69237">build#69237</a></summary>


| test_class                | test_method   | test_arguments                                              | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:--------------------------|:--------------|:------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| CloudArchiveRetentionTest | test_delete   | {"cloud_storage_type": 2, "retention_type": "retention.ms"} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69237#01981a57-58fa-4fc0-90a5-c2076b9d4d28 | FLAKY         | 119/120  | upstream reliability is '99.73821989528795'. current run reliability is '95.23809523809523'. drift is 4.50012 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69550">build#69550</a></summary>


| test_class                  | test_method                   | test_arguments                 | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:----------------------------|:------------------------------|:-------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| JavaCompressionTest         | test_upgrade_java_compression | {"compression_type": "gzip"}   | integration | https://buildkite.com/redpanda/redpanda/builds/69550#01983858-e73e-4307-ad53-b0ffe7de0166 | FLAKY         | 20/21    | upstream reliability is '95.1048951048951'. current run reliability is '95.23809523809523'. drift is -0.1332 and the allowed drift is set to 50. The test should PASS   |
| JavaCompressionTest         | test_upgrade_java_compression | {"compression_type": "gzip"}   | integration | https://buildkite.com/redpanda/redpanda/builds/69550#0198385f-7085-4006-80e0-0e7687081d73 | FLAKY         | 20/21    | upstream reliability is '95.1048951048951'. current run reliability is '95.23809523809523'. drift is -0.1332 and the allowed drift is set to 50. The test should PASS   |
| JavaCompressionTest         | test_upgrade_java_compression | {"compression_type": "lz4"}    | integration | https://buildkite.com/redpanda/redpanda/builds/69550#01983858-e73f-4eba-9d25-553f623334dd | FLAKY         | 20/21    | upstream reliability is '92.41379310344827'. current run reliability is '95.23809523809523'. drift is -2.8243 and the allowed drift is set to 50. The test should PASS  |
| JavaCompressionTest         | test_upgrade_java_compression | {"compression_type": "lz4"}    | integration | https://buildkite.com/redpanda/redpanda/builds/69550#0198385f-7087-4695-a11c-a27d0643b194 | FLAKY         | 20/21    | upstream reliability is '92.41379310344827'. current run reliability is '95.23809523809523'. drift is -2.8243 and the allowed drift is set to 50. The test should PASS  |
| JavaCompressionTest         | test_upgrade_java_compression | {"compression_type": "snappy"} | integration | https://buildkite.com/redpanda/redpanda/builds/69550#01983858-e740-4692-834b-1caced27a69d | FLAKY         | 20/21    | upstream reliability is '92.46575342465754'. current run reliability is '95.23809523809523'. drift is -2.77234 and the allowed drift is set to 50. The test should PASS |
| JavaCompressionTest         | test_upgrade_java_compression | {"compression_type": "snappy"} | integration | https://buildkite.com/redpanda/redpanda/builds/69550#0198385f-7088-49a3-9c67-b1c9c6b2f05f | FLAKY         | 20/21    | upstream reliability is '92.46575342465754'. current run reliability is '95.23809523809523'. drift is -2.77234 and the allowed drift is set to 50. The test should PASS |
| JavaCompressionTest         | test_upgrade_java_compression | {"compression_type": "zstd"}   | integration | https://buildkite.com/redpanda/redpanda/builds/69550#01983858-e741-4f0c-ac9b-1661dfeb75c8 | FLAKY         | 20/21    | upstream reliability is '92.3076923076923'. current run reliability is '95.23809523809523'. drift is -2.9304 and the allowed drift is set to 50. The test should PASS   |
| JavaCompressionTest         | test_upgrade_java_compression | {"compression_type": "zstd"}   | integration | https://buildkite.com/redpanda/redpanda/builds/69550#0198385f-708a-4441-bbff-30aa7891502f | FLAKY         | 20/21    | upstream reliability is '92.3076923076923'. current run reliability is '95.23809523809523'. drift is -2.9304 and the allowed drift is set to 50. The test should PASS   |
| TopicDeleteCloudStorageTest | drop_lifecycle_marker_test    | {"cloud_storage_type": 2}      | integration | https://buildkite.com/redpanda/redpanda/builds/69550#01983858-e739-4460-9a55-20c66f71d3b1 | FLAKY         | 3/21     | upstream reliability is '100.0'. current run reliability is '14.285714285714285'. drift is 85.71429 and the allowed drift is set to 50. The test should FAIL            |
| TopicDeleteCloudStorageTest | drop_lifecycle_marker_test    | {"cloud_storage_type": 2}      | integration | https://buildkite.com/redpanda/redpanda/builds/69550#0198385f-7087-4695-a11c-a27d0643b194 | FLAKY         | 3/21     | upstream reliability is '100.0'. current run reliability is '14.285714285714285'. drift is 85.71429 and the allowed drift is set to 50. The test should FAIL            |
| TopicDeleteCloudStorageTest | drop_lifecycle_marker_test    | {"cloud_storage_type": 1}      | integration | https://buildkite.com/redpanda/redpanda/builds/69550#01983858-e73b-4a20-9cd8-c91cf89ad001 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                  |
| TopicDeleteCloudStorageTest | drop_lifecycle_marker_test    | {"cloud_storage_type": 1}      | integration | https://buildkite.com/redpanda/redpanda/builds/69550#0198385f-7088-49a3-9c67-b1c9c6b2f05f | FLAKY         | 6/21     | upstream reliability is '100.0'. current run reliability is '28.57142857142857'. drift is 71.42857 and the allowed drift is set to 50. The test should FAIL             |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69573">build#69573</a></summary>


| test_class                  | test_method                | test_arguments            | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                       |
|:----------------------------|:---------------------------|:--------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------|
| TopicDeleteCloudStorageTest | drop_lifecycle_marker_test | {"cloud_storage_type": 2} | integration | https://buildkite.com/redpanda/redpanda/builds/69573#01983967-33cb-4b5d-bbca-f0e4251b691a | FLAKY         | 10/50    | upstream reliability is '100.0'. current run reliability is '14.893617021276595'. drift is 85.10638 and the allowed drift is set to 50. The test should FAIL |
| TopicDeleteCloudStorageTest | drop_lifecycle_marker_test | {"cloud_storage_type": 2} | integration | https://buildkite.com/redpanda/redpanda/builds/69573#01983967-33cb-4eb2-99f3-46f06a43d990 | FLAKY         | 10/50    | upstream reliability is '100.0'. current run reliability is '11.11111111111111'. drift is 88.88889 and the allowed drift is set to 50. The test should FAIL  |
| TopicDeleteCloudStorageTest | drop_lifecycle_marker_test | {"cloud_storage_type": 2} | integration | https://buildkite.com/redpanda/redpanda/builds/69573#01983967-33cb-4efa-ab5c-e358b3c8f954 | FLAKY         | 8/50     | upstream reliability is '100.0'. current run reliability is '4.545454545454546'. drift is 95.45455 and the allowed drift is set to 50. The test should FAIL  |
| TopicDeleteCloudStorageTest | drop_lifecycle_marker_test | {"cloud_storage_type": 2} | integration | https://buildkite.com/redpanda/redpanda/builds/69573#01983967-33d2-47ea-a27a-51ad85ba8423 | FLAKY         | 10/50    | upstream reliability is '100.0'. current run reliability is '11.11111111111111'. drift is 88.88889 and the allowed drift is set to 50. The test should FAIL  |
| TopicDeleteCloudStorageTest | drop_lifecycle_marker_test | {"cloud_storage_type": 2} | integration | https://buildkite.com/redpanda/redpanda/builds/69573#01983967-33d2-49ed-a66f-87b18de8850d | FLAKY         | 2/50     | The test has failed across all retries                                                                                                                       |
| TopicDeleteCloudStorageTest | drop_lifecycle_marker_test | {"cloud_storage_type": 2} | integration | https://buildkite.com/redpanda/redpanda/builds/69573#01983967-33d2-4e81-9e21-ad12fc30d7e2 | FLAKY         | 8/50     | upstream reliability is '100.0'. current run reliability is '8.695652173913043'. drift is 91.30435 and the allowed drift is set to 50. The test should FAIL  |
| TopicDeleteCloudStorageTest | drop_lifecycle_marker_test | {"cloud_storage_type": 1} | integration | https://buildkite.com/redpanda/redpanda/builds/69573#01983967-33cb-4b5d-bbca-f0e4251b691a | FLAKY         | 5/50     | upstream reliability is '100.0'. current run reliability is '4.25531914893617'. drift is 95.74468 and the allowed drift is set to 50. The test should FAIL   |
| TopicDeleteCloudStorageTest | drop_lifecycle_marker_test | {"cloud_storage_type": 1} | integration | https://buildkite.com/redpanda/redpanda/builds/69573#01983967-33cb-4eb2-99f3-46f06a43d990 | FLAKY         | 5/50     | upstream reliability is '100.0'. current run reliability is '4.25531914893617'. drift is 95.74468 and the allowed drift is set to 50. The test should FAIL   |
| TopicDeleteCloudStorageTest | drop_lifecycle_marker_test | {"cloud_storage_type": 1} | integration | https://buildkite.com/redpanda/redpanda/builds/69573#01983967-33cb-4efa-ab5c-e358b3c8f954 | FLAKY         | 4/50     | upstream reliability is '100.0'. current run reliability is '4.166666666666666'. drift is 95.83333 and the allowed drift is set to 50. The test should FAIL  |
| TopicDeleteCloudStorageTest | drop_lifecycle_marker_test | {"cloud_storage_type": 1} | integration | https://buildkite.com/redpanda/redpanda/builds/69573#01983967-33d2-47ea-a27a-51ad85ba8423 | FLAKY         | 9/50     | upstream reliability is '100.0'. current run reliability is '6.8181818181818175'. drift is 93.18182 and the allowed drift is set to 50. The test should FAIL |
| TopicDeleteCloudStorageTest | drop_lifecycle_marker_test | {"cloud_storage_type": 1} | integration | https://buildkite.com/redpanda/redpanda/builds/69573#01983967-33d2-49ed-a66f-87b18de8850d | FLAKY         | 5/50     | upstream reliability is '100.0'. current run reliability is '6.25'. drift is 93.75 and the allowed drift is set to 50. The test should FAIL                  |
| TopicDeleteCloudStorageTest | drop_lifecycle_marker_test | {"cloud_storage_type": 1} | integration | https://buildkite.com/redpanda/redpanda/builds/69573#01983967-33d2-4e81-9e21-ad12fc30d7e2 | FLAKY         | 3/50     | upstream reliability is '100.0'. current run reliability is '2.083333333333333'. drift is 97.91667 and the allowed drift is set to 50. The test should FAIL  |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69703">build#69703</a></summary>


| test_class                | test_method   | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:--------------------------|:--------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| EndToEndCloudTopicsTxTest | test_write    | null             | integration | https://buildkite.com/redpanda/redpanda/builds/69703#01984539-4dba-4d14-a566-b70eaa8b3d80 | FLAKY         | 18/21    | upstream reliability is '94.86301369863014'. current run reliability is '85.71428571428571'. drift is 9.14873 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

oleiman commented:
/ci-repeat 3
skip-redpanda-build
skip-units
dt-repeat=100
tests/rptest/tests/archive_retention_test.py::CloudArchiveRetentionTest.test_delete

oleiman committed:
dt/util: Introduce wait_until_with_progress_check

a wrapper around ducktape's wait_until that provides the ability to track
a crude approximation of progress

usage requires two timeouts:
  - timeout_sec: the overall timeout, after which the function will throw a
    TimeoutError no matter what
  - progress_sec: an incremental timeout, after each expiry we check that the
    value returned by a 'check' function has changed with respect to a cached
    result. if not, immediately throw TimeoutError

params:
  - check: the value we expect to change after each 'progress_sec'
  - condition: the condition we are waiting for. should be written in terms
    of the check function, e.g.
    check:
       def get_val():
           return admin.stuff()['val']
    condition:
       def condition(check: Callable):
           return check() > 10
       # used like
       condition(get_val)
  - timeout_sec (see above)
  - progress_sec (see above)
  - err_msg: Passed down to wait_until
  - logger: log progress after each progress_sec iteration

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman committed:
dt/archive: Use progress oriented wait for spillover uploads

This gives us a much longer timeout if block uploads grind to a halt but will
cut out if we stop making progress (i.e. uploading more spillover manifests)

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman commented:
/ci-repeat 3
skip-redpanda-build
skip-units
dt-repeat=200
tests/rptest/tests/archive_retention_test.py::CloudArchiveRetentionTest.test_delete

oleiman commented:
/ci-repeat 3
skip-redpanda-build
skip-units
dt-repeat=350
tests/rptest/tests/archive_retention_test.py::CloudArchiveRetentionTest.test_delete

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/69550-->
<!--188551a6-d546-4a70-a217-b96b48bf0a31-->
<!--c38db0e3-2ded-4b82-8871-b99a3dd7de1b-->
<!--1252f989-b363-4b88-9f55-3968a6cb9997-->
<!--b5fa8014-c133-46d1-8562-33b26f0ca88a-->
<!--title-->
## Retry command for [Build#69550](https://buildkite.com/redpanda/redpanda/builds/69550)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/topic_delete_test.py::TopicDeleteCloudStorageTest.drop_lifecycle_marker_test@{"cloud_storage_type":1}
tests/rptest/tests/topic_delete_test.py::TopicDeleteCloudStorageTest.drop_lifecycle_marker_test@{"cloud_storage_type":2}
```
<!--/body-->

oleiman commented:
/ci-repeat 3
skip-redpanda-build
skip-units
dt-repeat=30
tests/rptest/tests/topic_delete_test.py::TopicDeleteCloudStorageTest.drop_lifecycle_marker_test@{"cloud_storage_type":1}
tests/rptest/tests/topic_delete_test.py::TopicDeleteCloudStorageTest.drop_lifecycle_marker_test@{"cloud_storage_type":2}

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/69573-->
<!--faaaffcc-f8d2-43d1-afdc-727e767c0ef2-->
<!--113c69c4-2701-4e36-bce0-d94ed12f24b3-->
<!--f5575608-632f-4ff3-8ff1-58de05dc8c04-->
<!--d9124a8a-7909-42c7-a75c-812c384a8fa8-->
<!--054ddc01-c45f-4b87-8e38-4611ed34d75c-->
<!--2a229b89-a5a3-446b-801f-b1528e7616ac-->
<!--title-->
## Retry command for [Build#69573](https://buildkite.com/redpanda/redpanda/builds/69573)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/topic_delete_test.py::TopicDeleteCloudStorageTest.drop_lifecycle_marker_test@{"cloud_storage_type":1}
tests/rptest/tests/topic_delete_test.py::TopicDeleteCloudStorageTest.drop_lifecycle_marker_test@{"cloud_storage_type":2}
```
<!--/body-->

oleiman commented:
/ci-repeat 3
skip-redpanda-build
skip-units
dt-repeat=350
tests/rptest/tests/archive_retention_test.py::CloudArchiveRetentionTest.test_delete@{"cloud_storage_type":2,"retention_type":"retention.bytes"}

vbotbuildovich commented:
/backport v25.2.x

vbotbuildovich commented:
/backport v25.1.x

vbotbuildovich commented:
/backport v24.3.x

vbotbuildovich commented:
Failed to create a backport PR to v24.3.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-26883-v24.3.x-101 remotes/upstream/v24.3.x
git cherry-pick -x 0243b2d32d 4049767fca 2c520d18fc
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/16599554510).
